File: Day07/baekjoon-10799.py
def count_pieces(arrangement):
    stack = []
    count = 0
    
    for i in range(len(arrangement)):
        if arrangement[i] == '(':
            stack.append('(')
        else:  # ')'인 경우
            stack.pop()  # 스택에서 짝이 되는 '(' 제거
            
            # 레이저인 경우 (바로 이전 문자가 '('이면)
            if arrangement[i-1] == '(':
                count += len(stack)  # 현재 스택에 있는 막대기 수만큼 조각 추가
            else:  # 막대기의 끝인 경우
                count += 1  # 막대기가 끝날 때 조각 1개 추가
    
    return count

# 입력 받기
arrangement = input().strip()
print(count_pieces(arrangement))

File: Day07/baekjoon-10799.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <string>
using namespace std;
int count_pieces(vector<char> target)
{
    stack<char> stack;
    int count = 0;

    for (int i = 0; i < target.size(); i++)
    {
        if (target[i] == '(')
        {
            stack.push('(');
        }
        // ')' 인경우
        else
        {
            // 여태 '(' 만 넣다가 ')' 등장했으니 바로 () 짝이 만났다는것을 의미 => pop 수행
            stack.pop();

            // 레이저인 경우 (바로 이전 문자가 '(' 이면)
            if (target[i - 1] == '(')
            {
                count += stack.size();
            }
            else
            {
                count += 1;
            }
        }
    }
    return count;
}
int main(void)
{
    string s1;
    getline(cin, s1);
    vector<char> arrangement(s1.begin(), s1.end());
    cout << count_pieces(arrangement);
    return 0;
}